django/db/migrations/loader.py

```diff
@@ -210,6 +210,7 @@ def build_graph(self):
         self.replacements = {}
         for key, migration in self.disk_migrations.items():
             self.graph.add_node(key, migration)
+        for key, migration in self.disk_migrations.items():
```

Comment:
Suppose there are three migrations 0003 -> 0002 -> 0001, and migrations are loaded from disk in the order [0003, 0002, 0001].

Previously, Django was creating nodes and connections in this order: normal node 0003, dummy node 0002, dependency 0003 -> 0002, promoted node 0002 to normal node, dummy node 0001, dependency 0002 -> 0001, finally promoted dummy node 0001 to normal.

In this pull request, I am creating all nodes first, and then all dependencies: normal node 0003, normal node 0002, normal node 0001, dependency 0003 -> 0002, dependency 0002 -> 0001.

This is the reason for the extra "for": previously it was `for migration in migrations: add_node(migration); add_edges(migration)` now it's `for migration in migrations: add_node(migration); for migration in migrations: add_edges(migration)`.

Why the code still has dummy nodes? Because of squashing. A migration can depend on a squashed migration, and then DummyNode means a reference to a squashed migration. I didn't want to touch this though; I suspect this can be simplified, but it's already enough changes for a single pull request. Previously, DummyNodes were used for nodes that are to be constructed and for squashed migrations; now they're used only for squashed migrations and we can remove DummyNode.promote.

Reply:
Thanks for the explanation.

I think the code in lines 216/217 (`if migration.replaces: self.replacements[key] = migration`) should go into the first for loop. I believe adding nodes to the graph is closes related to keeping track of replacements than adding dependencies between nodes.

Can you please add the last paragraph of your comment in some form as a docstring to the `DummyNode` class.

Reply:
Agreed, done.